hibernate-core/src/test/java/org/hibernate/orm/test/query/criteria/CriteriaWindowFunctionTest.java

```diff
@@ -0,0 +1,282 @@
+					entity5.setTheInt( 5 );
+					entity5.setTheInteger( 5 );
+					entity5.setTheDouble( 9.0 );
+					entity5.setTheBoolean( false );
+					em.persist( entity5 );
+				}
+		);
+	}
+
+	@AfterEach
+	public void tearDown(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> session.createQuery( "delete from EntityOfBasics" ).executeUpdate()
+		);
+	}
+
+	@Test
+	public void testRowNumberWithoutOrder(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					HibernateCriteriaBuilder cb = session.getCriteriaBuilder();
+					CriteriaQuery<Long> cr = cb.createQuery( Long.class );
+					cr.from( EntityOfBasics.class );
+
+					JpaWindow window = cb.createWindow();
+					JpaExpression<Long> rowNumber = cb.rowNumber( window );
+
+					cr.select( rowNumber ).orderBy( cb.asc( cb.literal( 1 ) ) );
+					List<Long> resultList = session.createQuery( cr ).getResultList();
+					assertEquals( 5, resultList.size() );
+					assertEquals( 1L, resultList.get( 0 ) );
+					assertEquals( 2L, resultList.get( 1 ) );
+					assertEquals( 3L, resultList.get( 2 ) );
+					assertEquals( 4L, resultList.get( 3 ) );
+					assertEquals( 5L, resultList.get( 4 ) );
+				}
+		);
+	}
+
+	@Test
+	public void testFirstValue(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					HibernateCriteriaBuilder cb = session.getCriteriaBuilder();
+					CriteriaQuery<Integer> cr = cb.createQuery( Integer.class );
+					Root<EntityOfBasics> root = cr.from( EntityOfBasics.class );
+
+					JpaWindow window = cb.createWindow().orderBy( cb.desc( root.get( "theInt" ) ) );
+					JpaExpression<Integer> firstValue = cb.firstValue( root.get( "theInt" ), window );
+
+					cr.select( firstValue ).orderBy( cb.asc( cb.literal( 1 ) ) );
+					List<Integer> resultList = session.createQuery( cr ).getResultList();
+					assertEquals( 5, resultList.size() );
+					assertEquals( 13, resultList.get( 0 ) );
+					assertEquals( 13, resultList.get( 1 ) );
+					assertEquals( 13, resultList.get( 2 ) );
+					assertEquals( 13, resultList.get( 3 ) );
+					assertEquals( 13, resultList.get( 4 ) );
+				}
+		);
+	}
+
+	@Test
+	@SkipForDialect(dialectClass = SQLServerDialect.class)
+	@SkipForDialect(dialectClass = DB2Dialect.class)
+	public void testNthValue(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					HibernateCriteriaBuilder cb = session.getCriteriaBuilder();
+					CriteriaQuery<Integer> cr = cb.createQuery( Integer.class );
+					Root<EntityOfBasics> root = cr.from( EntityOfBasics.class );
+
+					JpaWindow window = cb.createWindow().orderBy( cb.desc( root.get( "theInt" ) ) );
+					JpaExpression<Integer> nthValue = cb.nthValue(
+							root.get( "theInt" ),
+							cb.literal( 2 ),
+							window
+					);
+
+					// todo marco : db2 throws java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
+					//  on getResultList() line, and SqlServer doesn't support nth_value function (even tho it's in Dialect)
```

Comment:
Here is a reference for the lack of the function in SQL Server: https://learn.microsoft.com/en-us/openspecs/sql_standards/ms-tsqliso02/16e25dcc-6f2c-4c15-9e71-1d8935c7a6ac

Not sure about DB2, but we can look into the stack trace together if you want.